zson/marshal.go

```diff
@@ -998,3 +1005,49 @@ func (u *UnmarshalZNGContext) lookupPrimitiveType(typ zng.Type) (reflect.Type, e
 	}
 	return reflect.TypeOf(v), nil
 }
+
+// MarshalReader provides a means to turn a sequence of Go values into
+// zng-marshaled records.  MarshalReader implements zbuf.Reader.  So, in
+// on goroutine, pull records from the zbuf.Reader with Read() and in
+// another supply values to the MarshalReader via the Supply() method.
+type MarshalReader struct {
+	ch        chan *zng.Record
+	done      chan error
+	marshaler *MarshalZNGContext
+}
+
+func NewMarshalReader(style TypeStyle) *MarshalReader {
+	m := NewZNGMarshaler()
+	m.Decorate(style)
+	return &MarshalReader{
+		ch:        make(chan *zng.Record),
+		done:      make(chan error),
+		marshaler: m,
+	}
+}
+
+func (m *MarshalReader) Close(err error) {
+	close(m.ch)
+	if err != nil {
+		m.done <- err
+	}
+	close(m.done)
+}
+
+func (m *MarshalReader) Supply(v interface{}) bool {
```

Comment:
Maybe change this to `func (m *MarshalReader) Encode(v interface{}) error` because, even though "encoder" isn't in the type name, it feels like https://pkg.go.dev/encoding/json@go1.16.2#Encoder.Encode?

Reply:
This pattern is different, e.g., I don't think it makes sense to send the same error to both sides but rather send the error to the reader and tell the scanner to stop.  Maybe the problem is with the name MarshalReader?  How about MarshalStream?